src/MonoTorrent/MonoTorrent.Streaming/LocalStream.cs

```diff
@@ -0,0 +1,155 @@
+// the following conditions:
+// 
+// The above copyright notice and this permission notice shall be
+// included in all copies or substantial portions of the Software.
+// 
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
+// LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+// WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+//
+
+
+using System;
+using System.Diagnostics;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+
+using MonoTorrent.Client;
+using MonoTorrent.Client.PiecePicking;
+
+namespace MonoTorrent.Streaming
+{
+    /// <summary>
+    /// A seekable Stream which can be used to access a <see cref="TorrentFile"/> while it is downloading.
+    /// If the stream seeks to a location which hasn't been downloaded yet, <see cref="Read(byte[], int, int)"/>
+    /// will block until the data is available. <see cref="ReadAsync(byte[], int, int, CancellationToken)"/>
+    /// will perform a non-blocking wait for the data.
+    /// </summary>
+    class LocalStream : Stream
+    {
+        long position;
+
+        public override bool CanRead => true;
+
+        public override bool CanSeek => true;
+
+        public override bool CanWrite => false;
+
+        public override long Length => File.Length;
+
+        public override long Position {
+            get => position;
+            set => Seek (value, SeekOrigin.Begin);
+        }
+
+        TorrentFile File { get; }
+
+        TorrentManager Manager { get; }
+
+        StreamingPiecePicker Picker { get; }
+
+        FileStream Stream { get; set; }
+
+        public LocalStream (TorrentManager manager, TorrentFile file, StreamingPiecePicker picker)
+        {
+            Manager = manager;
+            File = file;
+            Picker = picker;
+        }
+
+        protected override void Dispose (bool disposing)
+        {
+            base.Dispose (disposing);
+            if (disposing) {
+                Stream?.Dispose ();
+            }
+        }
+
+        public override void Flush ()
+            => throw new NotSupportedException ();
+
+        public override int Read (byte[] buffer, int offset, int count)
+            => ReadAsync (buffer, offset, count, CancellationToken.None).GetAwaiter ().GetResult ();
+
+        public override async Task<int> ReadAsync (byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+        {
+            // The torrent is treated as one big block of data, so this is the offset at which the current file's data starts at.
+            var torrentFileStartOffset = (long)File.StartPieceIndex * (long)Manager.Torrent.PieceLength + File.StartPieceOffset;
```

Comment:
`count` can be more then `Length - Position`. Moreover, there is a code in .net fw (i do not remember where) that passes buffer size as `count` and cycles while the result (bytes read count) is greater then 0. Also `Position` can be greater or equal then `Length`. You can check this things in the documentation, [here](https://docs.microsoft.com/en-us/dotnet/api/system.io.stream.position?view=netframework-4.8#remarks) and [here](https://docs.microsoft.com/en-us/dotnet/api/system.io.stream.read?view=netframework-4.8#System_IO_Stream_Read_System_Byte___System_Int32_System_Int32_). Here is [the example](https://github.com/ilyalatt/BitTorrentStorage/blob/master/src/BitTorrentStorage/BtsFileStream.cs#L111).

Reply:
It is typical for "count" to be equal to the buffer size, and that's how HttpStream itself is implemented. 

That said, I'm not sure what the issue is which led to your comment. Is there a mistake in the logic here?

Reply:
Sure. You can read a data of the next file because you do not check `count`. Also you can set `Position >= Length` and read the other file data. Ok, it is not true, because you call `Stream.ReadAsync` and it should handle all this cases. But you download the data for this stream using `startPiece` and `endPiece`. So there are cases when it waits for redundant data. Maybe forever :)

Reply:
Ah yes, I see what you mean!

If too much data were requested then it would either be fine or an exception would be thrown if the endPiece doesn't exist. 

With the current implementation it's not possible to read data from a second file because only one FileStream is open and you won't be able to over-read. 

If this is swapped to DiskManager then overreading could be an issue so clamping the offset/count to the appropriate level would be the right thing to do. That's a great point

Reply:
This is actually part of a problem! When the UWP media player plays a file, it over-reads on purpose but it also does not actually expect/require it's read to be fulfilled. For example if the file is 100kb in size UWP might request bytes 90 -> 110, but if you only give it bytes 90->100 it will work as expected.

So yes - you were right and I need to figure out a good mechanism for clamping the data :)